akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -247,3 +222,86 @@ private[akka] final class HeadOptionStage[T] extends SinkStage[T, Future[Option[
     }, p.future)
   }
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] class QueueSink[T]() extends SinkStage[T, SinkQueue[T]]("queueSink") {
+  type Requested[E] = Promise[Option[T]]
+
+  trait RequestElementCallback {
+    val requestElementFuture: Future[(Requested[T] ⇒ Unit)]
+  }
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    type Received[E] = Try[Option[E]]
+
+    val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+    require(maxBuffer > 0, "Buffer size must be greater than 0")
+
+    val buffer = FixedSizeBuffer[Received[T]](maxBuffer + 1)
+    var currentRequest: Option[Requested[T]] = None
+
+    val stageLogic = new GraphStageLogic(shape) with RequestElementCallback {
+      val requestElementPromise = Promise[(Requested[T] ⇒ Unit)]
+      val requestElementFuture = requestElementPromise.future
+
+      override def keepGoingAfterAllPortsClosed = true
+
+      override def preStart(): Unit = {
+        requestElementPromise.success(callback.invoke)
+        pull(in)
+      }
+
+      private val callback: AsyncCallback[Requested[T]] =
+        getAsyncCallback(promise ⇒
+          currentRequest match {
+            case Some(_) ⇒
+              promise.failure(new IllegalStateException("You have to wait for previous future to be resolved to send another request"))
+            case None ⇒
+              if (buffer.isEmpty) currentRequest = Some(promise)
+              else sendDownstream(promise)
+          })
+
+      def sendDownstream(promise: Requested[T]): Unit =
+        buffer.dequeue() match {
+          case Success(opt) ⇒ opt match {
+            case Some(e) ⇒ //emit element
+              promise.success(Some(e))
+            case None ⇒ //completed
+              promise.success(None)
+              completeStage()
+          }
+          case Failure(t) ⇒ //stream failed
+            promise.failure(t)
+            failStage(t)
+        }
+
+      def enqueueAndNotify(requested: Received[T]): Unit = {
+        buffer.enqueue(requested)
+        currentRequest match {
+          case Some(p) ⇒ sendDownstream(p)
+          case None    ⇒ //do nothing
+        }
+        currentRequest = None
```

Comment:
It’s clearer (and a tiny bit faster) to move this write into the Some-case above.